fluent-tests/src/test/java/com/azure/mgmttest/CompilationTests.java

```diff
@@ -170,4 +171,8 @@ public void testPolymophicSubClass() {
     public void testSchemaCleanup() {
         UserAssignedIdentity userAssignedIdentity = new UserAssignedIdentity();
     }
+
+    public void testResourceWithWritableName() {
+        FirewallRuleInner firewallRuleInner = new FirewallRuleInner();
```

Comment:
What would happen of the class, if we did not have this fix?

Reply:
diff after fix here: https://github.com/Azure/azure-sdk-for-java/pull/41476/commits/806b4e05d097f5c22f27deb047d64dc02c6e758c#diff-a6e36e0731c287838e46e86be1e06840866bec2766bbe888a0de3deac129ee80

1. `name` will be improperly shadowed in `FirewallRuleInner`, also getter improperly generated for it. 
  This is due to in current `getFieldPropeties`, we use `name` from `ProxyResource`, not the shadowed one in `ResourceWithWritableName`.  Now we start search from subclass to parent, to choose the one closest to the leaf.
  codegen fix [here](https://github.com/Azure/autorest.java/pull/2905/files#diff-9f20042b5cf02a7bf58023ee835bca8b87fd603b2304e10d68242142b3d61120R250-R251). 
2. setter won't be generated for the writable `name`. 
  Cause is the same.
3. fromJson would set `name` twice, once for `name` from `ProxyResource`, once for `name` from `ResourceWithWritableName`. 
  codegen fix [here](https://github.com/Azure/autorest.java/pull/2905/files#diff-9f20042b5cf02a7bf58023ee835bca8b87fd603b2304e10d68242142b3d61120R972-R1000)